### ISSUE TITLE
Streaming client: crash after unsafeEnd

### ISSUE THREAD

**Issue by srubin** [revdotcom/revai-node-sdk#60]
Calling RevAiStreamingClient's `client.unsafeEnd()` often leads to a crash:

```
Error [ERR_STREAM_PUSH_AFTER_EOF]: stream.push() after EOF
    at readableAddChunk (_stream_readable.js:257:32)
    at PassThrough.Readable.push (_stream_readable.js:224:10)
    at PassThrough.Transform.push (_stream_transform.js:151:32)
    at PassThrough.afterTransform (_stream_transform.js:92:10)
    at PassThrough._transform (_stream_passthrough.js:42:3)
    at PassThrough.Transform._read (_stream_transform.js:190:10)
    at PassThrough.Transform._write (_stream_transform.js:178:12)
    at doWrite (_stream_writable.js:415:12)
    at writeOrBuffer (_stream_writable.js:399:5)
    at PassThrough.Writable.write (_stream_writable.js:299:11)
Emitted 'error' event at:
    at errorOrDestroy (internal/streams/destroy.js:107:12)
    at readableAddChunk (_stream_readable.js:257:9)
    at PassThrough.Readable.push (_stream_readable.js:224:10)
    [... lines matching original stack trace ...]
    at writeOrBuffer (_stream_writable.js:399:5)
```

What's happening here is that `unsafeEnd` is calling `closeStreams`, which executes `this.responses.push(null)`. Once `null` has been pushed onto a stream, you cannot push to it again. Then, as the connection gets a response from rev.ai, we write to `this.response` again: https://github.com/revdotcom/revai-node-sdk/blob/ece2087801269f89e3cca8af7526ed7e6b356a0c/src/streaming-client.ts#L138

Reproduction of this crash: https://codesandbox.io/s/revai-node-sdk-crash-hpy8u?file=/src/index.js (assign a rev.ai API token to the "token" variable). This is the streaming example from the document + a call to `unsafeEnd` before we've sent/received all of the results.

Something like this should fix it:
```patch
diff --git a/src/streaming-client.ts b/src/streaming-client.ts
index 8374e49..f56715e 100644
--- a/src/streaming-client.ts
+++ b/src/streaming-client.ts
@@ -34,6 +34,7 @@ export class RevAiStreamingClient extends EventEmitter {
     private config: AudioConfig;
     private requests: PassThrough;
     private responses: PassThrough;
+    private streamsClosed: boolean;
 
     /**
      * @param accessToken Access token associated with the user's account
@@ -130,6 +131,9 @@ export class RevAiStreamingClient extends EventEmitter {
                 this.closeStreams();
             });
             connection.on('message', (message: any) => {
+                if (this.streamsClosed) {
+                    return;
+                }
                 if (message.type === 'utf8') {
                     let response = JSON.parse(message.utf8Data);
                     if ((response as StreamingResponse).type === 'connected') {
@@ -159,5 +163,6 @@ export class RevAiStreamingClient extends EventEmitter {
     private closeStreams(): void {
         this.requests.end();
         this.responses.push(null);
+        this.streamsClosed = true;
     }
-}
\ No newline at end of file
+}

```

**Comment by scottf-tvw** [comment]
I can confirm this issue as well. @srubin did you test that possible fix? This is important issue blocking when using ffmpeg piped to the stream duplex. especially when trying to avoid the 2hr limit while switching socket sessions and reassigning the pipe to the new duplex.

**Comment by kbridbur** [comment]
Thanks for filing the issue. We're working on a fix here: https://github.com/revdotcom/revai-node-sdk/pull/61

**Comment by scottf-tvw** [comment]
Something else to note, even if you close this with a flag the pipe may still have data enroute. 
You will need to be sure that the `BufferedDuplex` class in the `_write` method does not try to write to a closed passthrough. 
src/models/streaming/BufferDuplex.ts line 23

This will solve this issue
```
public _write(chunk: any, encoding: string, callback: any): boolean {
        if (this.input.writable){
            return this.input.write(chunk, encoding, callback);
        }
}
```
probably have to return false too if it's not true?

**Comment by srubin** [comment]
Confirmed that it is fixed in 2.6.1

**Comment by amikofalvy** [comment]
@scottf-tvw We are looking into your comment separately.
